### ISSUE TITLE
ContractCallException: Empty value (0x) returned from contract

### ISSUE THREAD

**Issue by IcyLemon** [web3j/web3j#550]
I encountered a  problem. 
My environment is：Centos 7 64，geth-1.7.3, web3j 3.3.1

It needs to be emphasized that I could be sure of that my contract was deployed and loaded normally, but once any of the APIs in the contract was invoked, the following exception(fragment) occurred:
`Caused by: org.web3j.tx.exceptions.ContractCallException: Empty value (0x) returned from contract
	at org.web3j.tx.Contract.executeCallSingleValueReturn(Contract.java:197)
	at org.web3j.tx.Contract.lambda$executeRemoteCallSingleValueReturn$1(Contract.java:251)
	at org.web3j.protocol.core.RemoteCall.send(RemoteCall.java:30)`

I want to know that what happened, because these APIs were all normal before, it seemed that this problem suddenly happened.

Who encountered this too? Thanks.

**Comment by glomarch** [comment]
public void deploy() throws Exception 
    {
        log.info("Deploying smart contract");
        String greetMessage = "Hi this is a greeter app";
        gasPrice=BigInteger.valueOf(100);
        
        this.contract = Greeter.deploy(web3j, credentials, gasPrice, gasLimit, greetMessage).send();
   
        log.info("contract deployed "  );
String gr = this.contract.greet().send();
    }
works fine.
however
log.info("loading contract");
    	this.contract = (Greeter) Greeter.load(
    			this.contactAddress, web3j, credentials, gasPrice, gasLimit);
String gr = this.contract.greet().send();
always return the above error.
Eventhough the contract binary returned from both contract are the same.
This is reproducible from many ethereum networks

**Comment by xhliu** [comment]
Try sleeping for a while after deploy(). I've noticed it's not synchronous.

**Comment by xiaods** [comment]
i came across this issue again:

```
chain-scrawler        | 2018-10-13 10:38:58.672 ERROR 1 --- [ool-22-thread-1] c.c.s.b.walletwatcher.EthereumWatcher    : Empty value (0x) returned from contract                  
chain-scrawler        | org.web3j.tx.exceptions.ContractCallException: Empty value (0x) returned from contract                       
chain-scrawler        |         at org.web3j.tx.Contract.executeCallSingleValueReturn(Contract.java:230)                                                                          
chain-scrawler        |         at org.web3j.tx.Contract.lambda$executeRemoteCallSingleValueReturn$1(Contract.java:297)   
chain-scrawler        |         at org.web3j.protocol.core.RemoteCall.send(RemoteCall.java:30)                                                                                    
chain-scrawler        |         at com.coin.scrawler.blockscan.walletwatcher.EthereumWatcher.getContract(EthereumWatcher.java:386)
```

**Comment by Bobavince** [comment]
Same on my side. Arch-based architecture, recurrent error on function's return (string, uint ...)

**Comment by Bobavince** [comment]
Don't know if it can help out, but I had the same issue and in my wrapped compiled solidity script, it was : 

    private static final String BINARY = "[{\"constant\":true,\"inputs\":[],\"name\":\"chars\",\"outputs\":[{\"name\":\"\",\"type\":\"string\"}],\"payable\":false,\"stateMutability\":\"view\",\"type\":\"function\"}, ( ... )se,\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"constructor\"}]";

instead of : 

    private static final String BINARY = "608060405234801561001057600080fd5b5060008054600160a060020a0319163317815560015561027f806100356000 (...) ffffffffffffffffffff163314156102515760005473ffffffffffffffffffffffffffffffffffffffff16ff5b5600a165627a7a72305820545e60010ee0d4b8f1315f9743499c370f058e1ab280aab90d9795cd074ff9e60029";

Which means, that web3j or solidity compiler does not correctly do its job. 
Personal issue, but it was not obvious.

**Comment by Straegge** [comment]
Issue still exists. Deploying the contract works fine, loading it does not, it is always invalid. You shouldn't have to deploy a new contract every time you run your program...is this issue known to the developers? Is it being fixed?

**Comment by metiago** [comment]
Hi guys,

I running a private go-ethereum on Azure.

I would like to add two scenarios that I'm getting issues using Web3j.

1. When I deploy a contract from a Java Application using Web3j API, which it seems to work fine, and then when I try to call contract's methods by Remix IDE it always return 0.

2. If I do the same and try to call the contract's methods from Java I get "Empty value (0x) returned from contract at xyz"

But, If I deploy the contract by REMIX IDE, both Java and REMIX work fine.

Is it a bug in the library ? Thanks

**Comment by AlexandrouR** [comment]
Is this still an issue with web3j 4.3.0

@glomarch in the code snippet above you do not set `this.contractAddress` after deploying, so you are not loading the contract correctly because `this.contractAddress` is `null`.

@Bobavince it seems that you have swapped the arguments to the flags of the web3j command line tools.

@metiago please open a separate issue if this is still a problem. It is unrelated to this issue. Please provide instructions to reproduce.

**Comment by mesudip** [comment]
Here's a sample abstract contract
```java
public class IERC20 extends Contract {
    private static final String BINARY = "";
    public static final String FUNC_APPROVE = "approve";
    public static final String FUNC_TOTALSUPPLY = "totalSupply";
    .....
}
```
What happens When you try to deploy this?
```java
// this will give no error.
IERC20 token = IERC20.deploy(web3j, adminCredentials, new DefaultGasProvider(), name,symbol);

// this will give the contract returned 0x error
token.getTotalSupply().send()
```

**How did we begin with a empty binary?**

`solc` generates both `.abi` and `.bin` files for abstract contracts. The `.abi` file will be blank for abstract contract. web3j will create wrapper out of it with blank `BINARY`

This can happen when you modify contract and accidently change a function to different signature.


**Resolution**
- The deploy method should throw an Exception when the BINARY field in contract is empty.
- We3j should generate only the `load` functions for abstract contracts not the `deploy` functions.

**Comment by AlexandrouR** [comment]
>  The `.abi` file will be blank 

do you mean the `.bin` will be blank?

I think a better resolution would be not to generate the `deploy` methods in this case.

**Comment by metiago** [comment]
Hi!
Any possible solution about this issue ? I'm getting the same error in Ubuntu 18 + Geth. 
`ERROR StatusLogger No log4j2 configuration file found. Using default configuration: logging only errors to the console.
Unlocked account: true
Exception in thread "main" org.web3j.tx.exceptions.ContractCallException: Empty value (0x) returned from contract
	at org.web3j.tx.Contract.executeCallSingleValueReturn(Contract.java:313)
	at org.web3j.tx.Contract.lambda$executeRemoteCallSingleValueReturn$1(Contract.java:397)
	at org.web3j.protocol.core.RemoteCall.send(RemoteCall.java:42)
	at io.tiago.pyg.contracts.DeployContract.defineMileage(DeployContract.java:56)
	at io.tiago.pyg.contracts.DeployContract.main(DeployContract.java:46)`

**Comment by zhanghuizong** [comment]
顶一下 , 同样遇到这个问题了

**Comment by it-wwh** [comment]
I ran into the same problem, but I solved it. The reason for my error was that I used the wrong RPC URL

**Comment by it-wwh** [comment]
I ran into the same problem, but I solved it. The reason for my error was that I used the wrong RPC URL

**Comment by saulgiordani** [comment]
Sorry to comment on a closed issue, but with web3j 4.9.4 i have the same issue. I'm writing integration tests and using OPEN_ETHEREUM as node. The deploy function correctly returns an instance of the java class, the contractaddress is not null, but when I call any function it returns 0x0. Loading the contract does not solve the issue.

With BESU node it works fine (but it gives other issues). Any suggestion?

**Comment by saideepak-lbg** [comment]
i am getting the issue with BESU node , got the contract address after deploy and getting the same error when a function is being called from java using wrapper classes
